integrations: Add webhook code, API endpoint and tests for Runscope.

XavierCooney opened the pull request:
As well as the integration, I also wrote a commit to allow the testing library to assert that a certain fixture doesn't send a message (relevant topic on chat.zulip.org about this: https://chat.zulip.org/#narrow/stream/development.20help/topic/Webhook.20Test.20Asserting.20No.20Response). Thanks in advance for any feedback.

aero31aero commented on zerver/webhooks/runscope/tests.py:
```diff
@@ -0,0 +1,65 @@
+from zerver.lib.test_classes import WebhookTestCase
+
+class RunscopeHookTests(WebhookTestCase):
+    STREAM_NAME = 'Runscope'
```
We prefer using lowercase stream names so it would be preferable to change this to `runscope`.

XavierCooney commented:
I've edited the commit history and also changed the `STREAM_NAME ` for the tests.

XavierCooney commented:
Yeah, the documentation wasn't part of the GCI task when I did this, but it makes sense to include it the same PR, I'll try and get it done soon.

eeshangarg commented on zerver/lib/test_classes.py:
```diff
@@ -581,6 +581,15 @@ def send_json_payload(self, user_profile: UserProfile, url: Text,
+        self.assert_json_success(result)
+        new_msg = self.get_last_message()
+        self.assertEqual(old_msg, new_msg)
+
```
I don't think you need this function! In order to see if a payload was ignored or not, see [this test in our GitLab integration](https://github.com/zulip/zulip/blob/master/zerver/webhooks/gitlab/tests.py#L378).

eeshangarg commented on zerver/lib/test_classes.py:
```diff
@@ -636,6 +645,13 @@ def send_and_test_stream_message(self, fixture_name: Text, expected_subject: Opt
+        if content_type is not None:
+            kwargs['content_type'] = content_type
+        self.send_json_payload_no_response(self.url, payload, **kwargs)
+
```
Again, you don't need this! See my comment above! :)

XavierCooney commented:
This is work in progress, I have to go right now, I'll finish this tomorrow, but I've added docs (not polished) and changed the relevant tests (I should have realised that I could've just patched from the test itself earlier, I actually updated the mypy annotations for those gitlab tests).

eeshangarg commented on zerver/webhooks/runscope/doc.md:
```diff
@@ -0,0 +1,27 @@
+Receive alerts on whenever your Runscope tests run!
```
Replace this with:
```
Get Zulip notifications for your Runscope tests!
```

eeshangarg commented on zerver/webhooks/runscope/doc.md:
```diff
@@ -0,0 +1,27 @@
+`{{api_url}}?api_key=abcdefgh&stream=Runscope%20Tests&topic=API%20Tests%201&only_on=fail`
+
+Now that you've got your webhook URL, go to your test settings, and add
+your URL to the webhook section:
```
If `test settings` and `webhook section` are actual physical locations on the Runscope website, make sure you properly bold and capitalise them like `**Webhook** section`. We bold the names of sections in third-party websites. The images you included don't seem to be getting rendered due to some issue with GitHub so I can't see what you are referring to, so this comment might not even make sense! :D

eeshangarg commented on zerver/webhooks/runscope/doc.md:
```diff
@@ -0,0 +1,27 @@
+
+![](/static/images/integrations/runscope/001.png)
+
+{!congratulations}
```
Replace with `{!congrats.md!}`

eeshangarg commented on zerver/webhooks/runscope/tests.py:
```diff
@@ -0,0 +1,71 @@
+    URL_TEMPLATE = "/api/v1/external/runscope?&api_key={api_key}&stream={stream}"
+    FIXTURE_DIR_NAME = 'runscope'
+
+    # Note: Include a test function per each distinct message condition your integration supports
```
I don't think you need this! :)

eeshangarg commented on zerver/webhooks/runscope/view.py:
```diff
@@ -0,0 +1,75 @@
+    scripts_passed = sum(request['scripts']['pass'] for request in payload['requests'])
+    scripts_executed = sum(request['scripts']['total'] for request in payload['requests'])
+
+    extraced_information = {
```
Minor typo: `extracted_information` :)

timabbott commented:
@XavierCooney are you planning to finish this?

XavierCooney commented:
If you can advise on the next step (I know I messed up with the merge commit, I tried to do it online, I'll clean it up) then I'll probably be able to finish this up a few days from now. Thanks

andersk commented:
Because the source repository for this pull request was deleted, this pull request has been automatically closed by GitHub as an unintended side effect of [renaming our default Git branch](https://chat.zulip.org/#narrow/stream/1-announce/topic/Git.20branch.20renamed.20to.20.E2.80.98main.E2.80.99). If this change is still desired, please open a new pull request, referencing this one in the description.
